core/shared/src/main/scala/org/http4s/util/SizedSeq.scala

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright 2013 http4s.org
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.http4s.util
+
+sealed trait SizedSeq[+A] {
+  def toSeq: Seq[A]
+}
+
+sealed abstract class SizedSeqBase[+A] extends SizedSeq[A] {
+  protected val seq: Seq[A]
+  override def toSeq: Seq[A] = seq
+}
+
+// format: off
+final case class SizedSeq0[+A]() extends SizedSeqBase[A] {
+  override val seq: Seq[A] = Seq.empty
+}
+
+final case class SizedSeq1[+A](s1: A) extends SizedSeqBase[A] {
+  override val seq: Seq[A] = Seq(s1)
+}
+
+final case class SizedSeq2[+A](s1: A, s2: A) extends SizedSeqBase[A] {
+  override val seq: Seq[A] = Seq(s1, s2)
+}
+
+final case class SizedSeq3[+A](s1: A, s2: A, s3: A) extends SizedSeqBase[A] {
+  override val seq: Seq[A] = Seq(s1, s2, s3)
+}
+
+final case class SizedSeq4[+A](s1: A, s2: A, s3: A, s4: A) extends SizedSeqBase[A] {
+  override val seq: Seq[A] = Seq(s1, s2, s3, s4)
+}
+
+final case class SizedSeq5[+A](s1: A, s2: A, s3: A, s4: A, s5: A) extends SizedSeqBase[A] {
+  override val seq: Seq[A] = Seq(s1, s2, s3, s4, s5)
+}
+
+final case class SizedSeq6[+A](s1: A, s2: A, s3: A, s4: A, s5: A, s6: A) extends SizedSeqBase[A] {
+  override val seq: Seq[A] = Seq(s1, s2, s3, s4, s5, s6)
+}
+
+final case class SizedSeq7[+A](s1: A, s2: A, s3: A, s4: A, s5: A, s6: A, s7: A) extends SizedSeqBase[A] {
+  override val seq: Seq[A] = Seq(s1, s2, s3, s4, s5, s6, s7)
+}
+
+final case class SizedSeq8[+A](s1: A, s2: A, s3: A, s4: A, s5: A, s6: A, s7: A, s8: A) extends SizedSeqBase[A] {
+  override val seq: Seq[A] = Seq(s1, s2, s3, s4, s5, s6, s7, s8)
+}
+
+final case class SizedSeq9[+A](s1: A, s2: A, s3: A, s4: A, s5: A, s6: A, s7: A, s8: A, s9: A) extends SizedSeqBase[A] {
+  override val seq: Seq[A] = Seq(s1, s2, s3, s4, s5, s6, s7, s8, s9)
+}
+
+final case class SizedSeq10[+A](s1: A, s2: A, s3: A, s4: A, s5: A, s6: A, s7: A, s8: A, s9: A, s10: A) extends SizedSeqBase[A] {
```

Comment:
10 is probably a practical upper bound here, but 22 is a traditional limit in Scala.  And because it's sealed, once it's merged, we're stuck with it until 1.0.  I might go to the absurdly large 22 while we're here.

Reply:
Added `SizedSeq11` to `SizedSeq22`.